lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/IMutableMetric.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using Org.Apache.REEF.Common.metrics.Api;
+
+namespace Org.Apache.REEF.Common.Metrics.MutableMetricsLayer
+{
+    internal interface IMutableMetric : IObserver<SnapshotRequest>
```

Comment:
> Could this also be an IObservable<MetricsType>? In other words: What's the design tradeoff between making this push out its updates vs. them being pulled from the outside?

I did not get this comment. Why do we need it to be IObservable<MetricsType> ?  Who is going to be observer?  Also I did not understand what this has to do with push vs. pull based mechanism.

Regarding push vs. pull, in push based mechanism we would need to have timers in each of the MetricSources than having a single timer in IMetricsSystem. IMetricsSystem will then be the observer to which these changes will be pushed in to. So the whole snapshot of the system is decentralized in some sense. Then the question arises when shall we push to sinks - should it be whenever we receive metrics from any source or again it should be governed by some timer in metrics system.

In pull based mechanism, we will have a single centralized control in metrics system. At some fixed time intervals it will pull metrics from registered sources and then push them in to the sinks. 

I would prefer a pull based mechanism if there are no strong arguments in favor of push.